src/libPMacc/include/preprocessor/struct.hpp

```diff
@@ -0,0 +1,207 @@
+    PMACC_PP_SELECT_TYPEID( 5,PMACC_PP_X_CREATE_VALUE_VARIABLE_WITH_PAREN, elem )
+
+#define PMACC_PP_X_CREATE_C_VALUE_VARIABLE(data,type,name,...) static const type name = __VA_ARGS__;
+#define PMACC_PP_CREATE_C_VALUE_VARIABLE(elem)                                 \
+    PMACC_PP_SELECT_TYPEID( 1,PMACC_PP_X_CREATE_C_VALUE_VARIABLE,elem )
+
+
+#define PMACC_PP_X1_INIT_VALUE_VARIABLE(data,type,name,...) (name(__VA_ARGS__))
+#define PMACC_PP_X_INIT_VALUE_VARIABLE(elem)                                   \
+    PMACC_PP_SELECT_TYPEID( 2,PMACC_PP_X1_INIT_VALUE_VARIABLE,elem )
+
+#define PMACC_PP_X_INIT_VALUE_VARIABLE_WITH_PAREN(elem)                        \
+    PMACC_PP_SELECT_TYPEID( 5,PMACC_PP_X1_INIT_VALUE_VARIABLE,elem )
+
+#define PMACC_PP_X_CREATE_C_STRING_VARIABLE(data,type,name,...) static const char* name;
+#define PMACC_PP_CREATE_C_STRING_VARIABLE(elem)                                \
+    PMACC_PP_SELECT_TYPEID( 3,PMACC_PP_X_CREATE_C_STRING_VARIABLE, elem )
+
+#define PMACC_PP_X_INIT_C_STRING_VARIABLE(data,type,name,...) const char* data::name = (char*)__VA_ARGS__;
+#define PMACC_PP_INIT_C_STRING_VARIABLE(elem)                                  \
+    PMACC_PP_SELECT_TYPEID( 3,PMACC_PP_X_INIT_C_STRING_VARIABLE,elem )
+
+#define PMACC_PP_X_CREATE_C_EXTENT(data,type,name,...) __VA_ARGS__
+#define PMACC_PP_CREATE_C_EXTENT(elem)                                         \
+    PMACC_PP_SELECT_TYPEID( 4,PMACC_PP_X_CREATE_C_EXTENT,elem )
+
+#define PMACC_PP_X1_ADD_DATA_TO_TYPEDESCRIPTION_MACRO(data,first,second) ((first,(data,PMACC_PP_REMOVE_PAREN(second))))
+#define PMACC_PP_X_ADD_DATA_TO_TYPEDESCRIPTION_MACRO(data,value) PMACC_PP_X1_ADD_DATA_TO_TYPEDESCRIPTION_MACRO(data,value)
+
+
+
+#define PMACC_PP_ADD_DATA_TO_TYPEDESCRIPTION_MACRO(r,data,elem)                \
+    PMACC_PP_X_ADD_DATA_TO_TYPEDESCRIPTION_MACRO(data,PMACC_PP_REMOVE_PAREN(elem))
+
+
+/** create constructor initialization of non static variables
+ *
+ * add an emty struct to the end of the sequences to avoid problems with empty sequences
+ *
+ * @param ... preprocessor sequence with macro types to inherit from
+ */
+#define PMACC_PP_INIT_VALUE_VARIABLES(op,...)                                  \
+    PMACC_PP_DEFER_REMOVE_PAREN() (                                            \
+        BOOST_PP_EXPAND(                                                       \
+          BOOST_PP_SEQ_TO_TUPLE (                                              \
+            BOOST_PP_SEQ_FOR_EACH(PMACC_PP_SEQ_MACRO_WITH_ACCESSOR,op,__VA_ARGS__((2,(a,b,EmptyStruct)))) \
+          )                                                                    \
+        )                                                                      \
+    )
+
+/** generate the definition of a struct
+ *
+ * @param namespace_name name of a unique namespace to avoid naming conflicts
+ * @param name name of the struct
+ * @param ... preprocessor sequence with type pairs
```

Comment:
```type pair``` is not really clear. This could easily be confused with a real pair of types e.g. a boost::mpl::pair.